pageserver/src/pgdatadir_mapping.rs

```diff
@@ -135,13 +141,21 @@ impl Timeline {
         let mut buf = self.get(key, lsn)?;
         let nblocks = buf.get_u32_le();
 
-        // Update relation size cache
-        self.update_cached_rel_size(tag, lsn, nblocks);
+        if latest {
+            // Update relation size cache only if "latest" flag is set.
+            // This flag is set by compute when it is working with most recent version of relation.
+            // Typically master compute node always set latest=true.
+            // Please notice, that even if compute node "by mistake" specifies old LSN but set
+            // latest=true, then it can not cause cache corruption, because with latest=true
+            // pageserver choose max(request_lsn, last_written_lsn) and so cached value will be
+            // associated with most recent value of LSN.
```

Comment:
There's still a race condition: `lsn` might be equal to `last_record_lsn` when we start processing the GetPage request, but more WAL could be processed, advancing `last_record_lsn`, before we get here.

Reply:
Sorry, still thinking that there is no race condition. 
Please notice that we compare LSN in `update_cached_rel_size` and update cached value only if new LSN is larger. If since the moment of start processing of GetPage request more WAL was generated which updates this relation, then cached is updated with larger LSN and GetPage will not change it. And if generated WAL doesn't affect this relation, then cached value is correct

Reply:
Hmm, ok. I'm still a bit nervous, but can't come up with any more issues :-).